README.rst
=================
 python-fs-stack
=================

``python-fs-stack`` provides a Python package that simplifies access to the
FamilySearch_ `REST-style API`_.

.. _FamilySearch: https://new.familysearch.org/
.. _REST-style API: https://devnet.familysearch.org/docs/api

:Home Page:
  http://pypi.python.org/pypi/python-fs-stack
:Source Code:
  https://github.com/familysearch-devnet/python-fs-stack


.. contents::


Dependencies
============

- Python 2.4 or later
- simplejson_, if using Python older than 2.6
- wsgi_intercept_ 0.5.0 or later (only required to run test suite)

.. _simplejson: http://pypi.python.org/pypi/simplejson
.. _wsgi_intercept: http://pypi.python.org/pypi/wsgi_intercept


Installation
============

Using pip_::

  pip install python-fs-stack

or using easy_install_ (from setuptools_ or distribute_)::

  easy_install python-fs-stack

or (after downloading manually)::

  python setup.py install

.. _pip: http://www.pip-installer.org/
.. _easy_install: http://peak.telecommunity.com/DevCenter/EasyInstall
.. _setuptools: http://pypi.python.org/pypi/setuptools
.. _distribute: http://pypi.python.org/pypi/distribute


Example Usage
=============

First, import the FamilySearch class::

  from familysearch import FamilySearch


Authenticating with FamilySearch
--------------------------------

``python-fs-stack`` supports several ways of initiating a session with
FamilySearch, including Basic Authentication, OAuth, and resuming a previous
session.

Log in immediately with Basic Authentication::

  fs = FamilySearch('ClientApp/1.0', 'developer_key', 'username', 'password')

Log in in a separate step with Basic Authentication::

  fs = FamilySearch('ClientApp/1.0', 'developer_key')
  fs.login('username', 'password')

Log in in two steps with Basic Authentication::

  fs = FamilySearch('ClientApp/1.0', 'developer_key')
  fs.initialize()
  fs.authenticate('username', 'password')

Log in with OAuth::

  import webbrowser
  fs = FamilySearch('ClientApp/1.0', 'developer_key')
  fs.request_token()
  webbrowser.open(fs.authorize())
  # [Enter username and password into browser window that opens]
  verifier = [verifier from resulting web page]
  fs.access_token(verifier)

Resume a previous session::

  fs = FamilySearch('ClientApp/1.0', 'developer_key', session='session_id')

Use the production system instead of the reference system::

  fs = FamilySearch('ClientApp/1.0', 'developer_key', base='https://api.familysearch.org')


Maintaining and Ending a Session
--------------------------------

Keep the current session active::

  fs.session()

Log out::

  fs.logout()


Accessing Family Tree Information
---------------------------------

Print current user's family tree details::

  print fs.person()

To specify a person ID to retrieve, pass the ID as an argument::

  print fs.person('ABCD-123')

To print multiple family tree entries, pass a list of IDs as an argument. To
pass additional parameters to the API, simply pass them as named arguments::

  print fs.person(['ABCD-123', 'EFGH-456'], events='all', children='all')

Print the latest version of a list of persons (this request is more lightweight
than a full person request, so it supports more IDs at once)::

  for person in fs.version(['ABCD-123', 'EFGH-456']):
      print person['id'], person['version']

Print the contents of a persona::

  print fs.persona('ABCD-123')

Print current user's pedigree::

  print fs.pedigree()

Format the pedigree output more nicely::

  import pprint
  pprint.pprint(fs.pedigree())


Searching for Persons in the Family Tree
----------------------------------------

Search for a male named John Smith::

  results = fs.search(givenName='John', familyName='Smith', gender='Male', maxResults=10)

Retrieve the second page of the previous search::

  more_results = fs.search(contextId=results[0]['contextId'], maxResults=10, startIndex=10)

Search for an exact match for John Smith (use an ``options`` dict to specify
options with periods in their names)::

  results = fs.search(options={'givenName.exact': 'John', 'familyName.exact': 'Smith'}, gender='Male', maxResults=10)


Searching for Possible Duplicates
---------------------------------

Search for possible duplicates of a person::

  matches = fs.match('ABCD-123')

Compute match score between two persons::

  match = fs.match('ABCD-123', id='EFGH-456')

Search for possible duplicates matching specified parameters::

  matches = fs.match(givenName='John', familyName='Smith', gender='Male', birthDate='1900', birthPlace='USA', deathDate='1950', deathPlace='USA')


Standardizing Places, Names, and Dates
--------------------------------------

Look up a place by name::

  place = fs.place(place='paris')

Look up a place by ID::

  place = fs.place(5061509)

Look up a list of places by ID::

  places = fs.place([5061509, 5061446])

Look up a place by name, showing only the most likely result, returning results in another locale::

  place = fs.place(place='germany', filter=True, locale='de')

Standardize a name::

  name = fs.name('John Smith')

Standardize a list of names::

  names = fs.name(['John Smith', 'Jane Doe'])

Standardize a date::

  date = fs.date('1-1-11')

Standardize a list of dates::

  dates = fs.date(['1-1-11', 'december 31 1999'])


CHANGELOG.rst
===========================
 python-fs-stack Changelog
===========================

0.3
---


0.2 (8 Jun 2011)
----------------

* Supports Python 2.4
* Supports Authorities requests (place, name, date, and culture)
* Supports Person Version Read and Persona Read
* Properly sends session ID on first request without needing to call session()
* Search and Match return a single object instead of a one-item list
* Supports pickling using any protocol (0-2)
* Allows authorize() to supply arbitrary parameters
* Handles HTTP 401 (Unauthorized) for Python 2.4 and 2.5 (GitHub issue 1)
* Full regression test coverage (requires wsgi_intercept)
* The login.py example reads the password more portably


0.1 (21 Jan 2011)
-----------------

* Initial release
* Supports Family Tree Read requests only (person, pedigree, search, match)
* Returns parsed JSON
* Requires Python 2.5


familysearch/__init__.py
"""
A library to interact with the FamilySearch API

Licensed under the FamilySearch API License Agreement;
see the included LICENSE file for details.

Example usage:

from familysearch import FamilySearch

# Log in immediately with Basic Authentication
fs = FamilySearch('ClientApp/1.0', 'developer_key', 'username', 'password')

# Log in in a separate step with Basic Authentication
fs = FamilySearch('ClientApp/1.0', 'developer_key')
fs.login('username', 'password')

# Log in in two steps with Basic Authentication
fs = FamilySearch('ClientApp/1.0', 'developer_key')
fs.initialize()
fs.authenticate('username', 'password')

# Resume a previous session
fs = FamilySearch('ClientApp/1.0', 'developer_key', session='session_id')

# Use the production system instead of the reference system
fs = FamilySearch('ClientApp/1.0', 'developer_key', base='https://api.familysearch.org')

# Log in with OAuth
import webbrowser
fs = FamilySearch('ClientApp/1.0', 'developer_key')
fs.request_token()
webbrowser.open(fs.authorize())
# [Enter username and password into browser window that opens]
verifier = [verifier from resulting web page]
fs.access_token(verifier)

# Keep current session active
fs.session()

# Log out
fs.logout()

# Retrieve current user's family tree details
me = fs.person()

# Retrieve family tree details for a person or several persons
person = fs.person('ABCD-123')
persons = fs.person(['ABCD-123', 'EFGH-456'])

# Retrieve family tree details for a list of persons, specifying parameters
persons = fs.person(['ABCD-123', 'EFGH-456'], events='all', children='all')

# Retrieve current user's pedigree
pedigree =  fs.pedigree()

# Search for a male named John Smith and retrieve the second page
results = fs.search(givenName='John', familyName='Smith', gender='Male', maxResults=10)
more_results = fs.search(contextId=results[0]['contextId'], maxResults=10, startIndex=10)

# Search for possible duplicates of a person
matches = fs.match('ABCD-123')

# Compute match score between two persons
match = fs.match('ABCD-123', id='EFGH-456')

# For more examples, see README.rst
"""

import urllib
import urllib2
import urlparse

# Support Python < 2.6
if not hasattr(urlparse, 'parse_qs'):
    import cgi
    urlparse.parse_qs = cgi.parse_qs

__version__ = '0.3pre'


class object(object): pass
class FamilySearch(object):

    """
    A FamilySearch API proxy

    The constructor must be called with a user-agent string and a developer key.
    A username, password, session ID, and base URL are all optional.

    Public methods:

    login -- log into FamilySearch with a username and password
    initialize -- create an unauthenticated session
    authenticate -- authenticate a session with a username and password
    logout -- log out of FamilySearch, terminating the current session
    session -- keep current session active
    request_token -- get an OAuth request token
    authorize -- construct OAuth authorization URL
    access_token -- get an OAuth access token (to complete the login process)

    person -- get a person or list of persons from the family tree
    persona -- get a persona or list of personas from the family tree
    version -- get the latest version number of a person from the family tree
    pedigree -- get the pedigree of a person or list of persons
    search -- search for persons in the family tree
    match -- search for possible duplicates in the family tree

    place -- standardize a place name
    name -- standardize a person name
    date -- standardize a date
    culture -- look up culture IDs

    Public attributes:

    logged_in -- flag indicating whether this proxy instance is logged in
    """

    def __init__(self, agent, key, username=None, password=None, session=None,
                 base='http://www.dev.usys.org'):
        """
        Instantiate a FamilySearch proxy object.

        Keyword arguments:
        agent -- User-agent string to use for requests
        key -- FamilySearch developer key (optional if reusing an existing session ID)
        username (optional)
        password (optional)
        session (optional) -- existing session ID to reuse
        base (optional) -- base URL for the API;
                           defaults to 'http://www.dev.usys.org' (the Reference System)
        """
        self.agent = '%s Python-FS-Stack/%s' % (agent, __version__)
        self.key = key
        self.session_id = session
        self.base = base
        self.opener = urllib2.build_opener()

        for mixin in self.__class__.__bases__:
            mixin.__init__(self)

        if username and password:
            self.login(username, password)

    def __getstate__(self):
        """
        Return a tuple containing the state necessary to pickle this instance.
        """
        return ({'agent': ' '.join(self.agent.split(' ')[:-1]),
                 'key': self.key,
                 'session': self.session_id,
                 'base': self.base},
                dict([(session, secret)
                      for (session, secret)
                      in self.oauth_secrets.iteritems()
                      if session == self.session_id]))

    def __setstate__(self, state):
        """
        Restore the saved state obtained from unpickling this instance.
        """
        self.__init__(**state[0])
        self.oauth_secrets = state[1]
        if self.session_id in self.oauth_secrets:
            self.logged_in = False

    def _request(self, url, data=None):
        """
        Make a GET or a POST request to the FamilySearch API.

        Adds the User-Agent header and sets the response format to JSON.
        If the data argument is supplied, makes a POST request.
        Returns a file-like object representing the response.

        """
        url = self._add_json_format(url)
        if self.logged_in and not self.cookies:
            # Add sessionId parameter to url if cookie is not set
            url = self._add_query_params(url, sessionId=self.session_id)
        request = urllib2.Request(url, data)
        request.add_header('User-Agent', self.agent)
        try:
            return self.opener.open(request)
        except urllib2.HTTPError, error:
            if error.code == 401:
                self.logged_in = False
            raise

    def _add_subpath(self, url, subpath):
        """
        Add a subpath to the path component of the given URL.

        For example, adding sub to http://example.com/path?query
        becomes http://example.com/path/sub?query.

        """
        parts = urlparse.urlsplit(url)
        path = parts[2] + '/' + subpath
        return urlparse.urlunsplit((parts[0], parts[1], path, parts[3], parts[4]))

    def _add_query_params(self, url, params={}, **kw_params):
        """
        Add the specified query parameters to the given URL.

        Parameters can be passed either as a dictionary or as keyword arguments.

        """
        parts = urlparse.urlsplit(url)
        query_parts = urlparse.parse_qs(parts[3])
        query_parts.update(params)
        query_parts.update(kw_params)
        query = urllib.urlencode(query_parts, True)
        return urlparse.urlunsplit((parts[0], parts[1], parts[2], query, parts[4]))

    def _add_json_format(self, url):
        """
        Add dataFormat=application/json to the query string of the given URL.
        """
        return self._add_query_params(url, dataFormat='application/json')

import identity_v2
import familytree_v2
import authorities_v1
